includes/class-sensei-course-manual-enrolment-provider.php

```diff
@@ -0,0 +1,242 @@
+
+		$migration_log      = [];
+		$is_legacy_enrolled = ! empty( $course_progress_comment_id );
+
+		$migration_log['had_progress'] = $is_legacy_enrolled;
+
+		/**
+		 * Allows other providers to have an opinion about whether or not a user was enrolled pre-3.0.0.
+		 * This should only be called once per user/course just after upgrading from a pre-3.0.0 version of Sensei.
+		 *
+		 * @since 3.0.0
+		 *
+		 * @param bool $is_legacy_enrolled If the user was actually enrolled before 3.0.0 migration.
+		 * @param int  $user_id            User ID.
+		 * @param int  $course_id          Course post ID.
+		 */
+		$is_legacy_enrolled = apply_filters( 'sensei_is_legacy_enrolled', $is_legacy_enrolled, $user_id, $course_id );
+
+		$migration_log['is_enrolled'] = $is_legacy_enrolled;
+
+		if ( $is_legacy_enrolled ) {
+			$this->add_student_enrolment( $user_id, $course_id );
+		}
+
+		$this->set_migrated_legacy_enrolment_status( $user_id, $course_id, $migration_log );
+	}
+
+	/**
+	 * Check if a user's enrolment status needs to be migrated from pre-3.0.0 enrolment.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return bool
+	 */
+	private function needs_legacy_migration( $user_id, $course_id ) {
+		if ( ! get_option( 'sensei_enrolment_legacy' ) ) {
+			return false;
+		}
+
+		return ! $this->has_migrated_legacy_enrolment( $user_id, $course_id );
+	}
+
+	/**
+	 * Checks if we've migrated legacy enrolment status.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return bool
+	 */
+	private function has_migrated_legacy_enrolment( $user_id, $course_id ) {
+		$term          = Sensei_Learner::get_learner_term( $user_id );
+		$migration_log = get_term_meta( $term->term_id, $this->get_legacy_migration_status_meta_key( $course_id ), true );
+
+		return ! empty( $migration_log );
+	}
+
+	/**
+	 * Update legacy migration status.
+	 *
+	 * @param int   $user_id       User ID.
+	 * @param int   $course_id     Course post ID.
+	 * @param array $migration_log Log of the migration.
+	 */
+	private function set_migrated_legacy_enrolment_status( $user_id, $course_id, $migration_log ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+		update_term_meta( $term->term_id, $this->get_legacy_migration_status_meta_key( $course_id ), wp_json_encode( $migration_log ) );
+	}
+
+	/**
+	 * Gets the meta key used to track legacy enrolment status migration.
+	 *
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return string
+	 */
+	private function get_legacy_migration_status_meta_key( $course_id ) {
+		return self::META_PREFIX_LEGACY_MIGRATION . $course_id;
+	}
+
+	/**
+	 * Gets the version of the enrolment provider logic. If this changes, enrolment will be recalculated.
+	 *
+	 * @return int
+	 */
+	public static function get_version() {
+		// @todo change this to just increment an integer.
+		return filemtime( __FILE__ );
```

Comment:
Just adding this as a comment to be resolved before merging 🙂

Reply:
I think we should keep this in until release, but I'll add a task to create a PR when this is merged.